File: cpp/src/Source.cpp
#include <Source.h>

#include <string>
#include <limits>
#include <vector>

// JSON Keys
#define AGENCYID_KEY "AgencyID"
#define AUTHOR_KEY "Author"
#define TYPE_KEY "Type"

namespace processingformats {
Source::Source() {
	agencyId = "";
	author = "";
	type = "";
}

Source::Source(std::string newAgencyId, std::string newAuthor,
		std::string newType) {
	agencyId = newAgencyId;
	author = newAuthor;
	type = newType;
}

Source::Source(rapidjson::Value &json) {
	// required values
	// agencyId
	if ((json.HasMember(AGENCYID_KEY) == true)
			&& (json[AGENCYID_KEY].IsString() == true)) {
		agencyId = std::string(json[AGENCYID_KEY].GetString(),
								json[AGENCYID_KEY].GetStringLength());
	} else {
		agencyId = "";
	}

	// author
	if ((json.HasMember(AUTHOR_KEY) == true)
			&& (json[AUTHOR_KEY].IsString() == true)) {
		author = std::string(json[AUTHOR_KEY].GetString(),
								json[AUTHOR_KEY].GetStringLength());
	} else {
		author = "";
	}

	// type
	if ((json.HasMember(TYPE_KEY) == true)
			&& (json[TYPE_KEY].IsString() == true)) {
		type = std::string(json[TYPE_KEY].GetString(),
								json[TYPE_KEY].GetStringLength());
	} else {
		type = "";
	}
}

Source::Source(const Source & newSource) {
	agencyId = newSource.agencyId;
	author = newSource.author;
	type = newSource.type;
}

Source::~Source() {
	agencyId = "";
	author = "";
	type = "";
}

rapidjson::Value & Source::toJSON(
		rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// agencyId
	if (agencyId != "") {
		rapidjson::Value agencyIdValue;
		agencyIdValue.SetString(rapidjson::StringRef(agencyId.c_str()),
								allocator);
		json.AddMember(AGENCYID_KEY, agencyIdValue, allocator);
	}

	// author
	if (author != "") {
		rapidjson::Value authorValue;
		authorValue.SetString(rapidjson::StringRef(author.c_str()), allocator);
		json.AddMember(AUTHOR_KEY, authorValue, allocator);
	}

	// type
	if (type != "") {
		rapidjson::Value typeValue;
		typeValue.SetString(rapidjson::StringRef(type.c_str()), allocator);
		json.AddMember(TYPE_KEY, typeValue, allocator);
	}

	return (json);
}

std::vector<std::string> Source::getErrors() {
	std::vector<std::string> errorlist;

	// agencyId
	if (agencyId == "") {
		// empty agencyid
		errorlist.push_back("Empty AgencyID in Source class.");
	}

	// author
	if (author == "") {
		// empty author
		errorlist.push_back("Empty Author in Source class.");
	}

	// since agencyid and author are free text strings, no further validation is
	// required for them

	// type
	if (type == "") {
		// empty type
		errorlist.push_back("Empty Type in Source class.");
	} else {
		bool match = false;
		if (type == "Unknown") {
			match = true;
		} else if (type == "LocalHuman") {
			match = true;
		} else if (type == "LocalAutomatic") {
			match = true;
		} else if (type == "ContributedHuman") {
			match = true;
		} else if (type == "ContributedAutomatic") {
			match = true;
		} else {
			match = false;
		}

		if (!match) {
			// invalid type
			errorlist.push_back("Invalid Type in Source Class.");
		}
	}

	// return the list of errors
	return (errorlist);
}

bool Source::isEmpty() {
	if (agencyId != "") {
		return (false);
	}
	if (author != "") {
		return (false);
	}
	if (type != "") {
		return (false);
	}

	return (true);
}
}  // namespace processingformats

File: source.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#stdlib imports
import json

class Source:
    """
    A conversion class used to create, parse, and validate source data as part
    of processing data.
    """
    
    #JSON Keys
    AGENCYID_KEY = "AgencyID" #required
    AUTHOR_KEY = "Author" #required
    TYPE_KEY = "Type" #required
    
    #Intialize source object
    def __init__ (self, newAgencyID = None, newAuthor = None, newType = None):
        
        if newAgencyID is not None:
            self.agencyID = newAgencyID
            
        if newAuthor is not None:
            self.author = newAuthor
            
        if newType is not None:
            self.type = newType
            
    #Populate object from JSON formatted string
    def fromJSONString (self, JSONString):
        JSONObject = json.loads(JSONString)
        self.fromDict(JSONObject)
        
    #Populates object from a dictionary
    def fromDict(self, aDict):
        
        try:
            self.agencyID = aDict[self.AGENCYID_KEY]
            self.author = aDict[self.AUTHOR_KEY]
            self.type = aDict[self.TYPE_KEY]
            
        except(ValueError, KeyError, TypeError) as e:
            print ("Dictionary format error, missing required keys: %s" % e)
            
            
    #Converts object to JSON formatted string
    def toJSONString(self):
        JSONObect = self.toDict()
        
        return json.dumps(JSONObject, ensure_ascii=False)
    
    #Converts the object to a dictionary
    def toDict(self):
        
        aDict = {}
        
        try:
            aDict[self.AGENCYID_KEY] = self.agencyID
            aDict[self.AUTHOR_KEY] = self.author
            aDict[self.TYPE_KEY] = self.type
            
        except(NameError, AttributeError) as e:
            print("Missing required data error: %s" % e)
            
            
        return aDict
    
    #Checks to see if object is valid
    def isValid(self):
        errorList = self.getErrors()
        
        return not errorList
    
    #Gets a list of object validation errors
    def getErrors(self):
        
        errorList = []
        
        #AgencyID
        try:
            if self.agencyID == '':
                errorList.append('Empty AgencyID in Source Class')
                
        except(NameError, AttributeError):
            errorList.append('No AgencyID in Source Class')
            
        #Author
        try:
            if self.author == '':
                errorList.append('Empty Author in Source Class')
                
        except(NameError, AttributeError):
            errorList.append('No Author in Source Class')
            
        #Type
        try:
            match = false
            
            if self.type == '':
                errorList.append('Empty Type in Source Class')
            
            elif self.type == 'Unkown':
                match = true
                    
            elif self.type == 'LocalHuman':
                match = true
                    
            elif self.type == 'LocalAutomatic':
                match = true
    
            elif self.type == 'ContributedHuman':
                match = true
                
            elif self.type == 'ContributedAutomatic':
                match = true
                
            else:
                match = false
                
            if not match:
                errorList.append('Invalid Type in Source Class')
                    
        except(NameError, AttributeError):
            errorList.append('No Type in Source Class')
                    
        
        return errorList
    
